crates/turborepo-telemetry/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+        Some(s) => {
+            let result = s.send(event);
+            if let Err(err) = result {
+                debug!("failed to send telemetry event. error: {}", err)
+            }
+        }
+        None => {
+            if cfg!(debug_assertions) {
+                error!("[DEVELOPMENT ERROR] telemetry sender not initialized");
+                panic!();
+            }
+            debug!("telemetry sender not initialized");
+        }
+    }
+}
+
+/// Starts the `Worker` on a separate tokio thread. Returns an `TelemetrySender`
+/// and an `TelemetryHandle`.
+///
+/// We have two different types because the TelemetrySender should be shared
+/// across threads (i.e. Clone + Send), while the TelemetryHandle cannot be
+/// shared since it contains the structs necessary to shut down the worker.
+pub fn init_telemetry(
+    client: impl telemetry::TelemetryClient + Clone + Send + Sync + 'static,
+    ui: UI,
+) -> Result<TelemetryHandle, Error> {
+    // make sure we're not already initialized
+    if SENDER_INSTANCE.get().is_some() {
+        debug!("telemetry already initialized");
+        return Err(Error::AlreadyInitialized());
+    }
+
+    let (tx, rx) = mpsc::unbounded_channel();
+    let (cancel_tx, cancel_rx) = oneshot::channel();
+    let mut config = TelemetryConfig::new(ui)?;
+    config.show_alert();
+
+    let session_id = Uuid::new_v4();
+    let worker = Worker {
+        rx,
+        buffer: Vec::new(),
+        senders: FuturesUnordered::new(),
+        exit_ch: cancel_tx,
+        client,
+        session_id: session_id.to_string(),
+        telemetry_id: config.get_id().to_string(),
+        enabled: config.is_enabled(),
+        ui,
+    };
+    let handle = worker.start();
+
+    let telemetry_handle = TelemetryHandle {
+        exit_ch: cancel_rx,
+        handle,
+    };
+
+    // track the sender as global to avoid passing it around
+    SENDER_INSTANCE.set(tx).unwrap();
+
+    // return
+    Ok(telemetry_handle)
+}
+
+impl TelemetryHandle {
+    async fn close(self) -> Result<(), Error> {
+        drop(self.exit_ch);
+        self.handle.await?;
+
+        Ok(())
+    }
+
+    /// Closes the handle with an explicit timeout. If the handle fails to close
+    /// within that timeout, it will log an error and drop the handle.
+    pub async fn close_with_timeout(self) {
+        if let Err(err) = tokio::time::timeout(EVENT_TIMEOUT, self.close()).await {
+            debug!("failed to close telemetry handle. error: {}", err)
+        } else {
+            debug!("telemetry handle closed")
+        }
+    }
+}
+
+struct Worker<C> {
+    rx: mpsc::UnboundedReceiver<TelemetryEvent>,
+    buffer: Vec<TelemetryEvent>,
+    senders: FuturesUnordered<JoinHandle<()>>,
+    // Used to cancel the worker
+    exit_ch: oneshot::Sender<()>,
+    client: C,
+    telemetry_id: String,
+    session_id: String,
+    enabled: bool,
+    ui: UI,
+}
+
+impl<C: telemetry::TelemetryClient + Clone + Send + Sync + 'static> Worker<C> {
```

Comment:
How much is the code different between here and analytics? Because if it's very similar, could we combine events and use the same worker?

Reply:
I explored this a little, and it could be made to work with multiple clients and event queues. My main concern was that the telemetry crate needs the worker to be spawned to work correctly and bundling it with the analytics worker felt a little strange (I split it out to an event-worker) crate. One of the initial goals was to make the telemetry crate as modular as possible, so shipping the worker as part of the crate feels right. But I could be convinced otherwise.

Reply:
I chatted more about this with @gsoltis and @Zertsov during our project sync and we're leaning more towards keeping the workers separate because the tasks that these workers are concerned with (analytics and telemetry) have different requirements. 

We _need_ analytics data to support product features (time saved, usage dashboads etc.) and thus it should have different SLA's around this data. 

For telemetry, the more data we get the better, but we shouldn't spend an excessive amount of time blocking on shutdown to send it, and we shouldn't care if some of the data is lost. 

We also still want to keep telemetry as self contained as possible moving forward.